### ISSUE TITLE
is there also optimization taking place ?

### ISSUE THREAD

**Issue by YitzhakSp** [nicholas-leonard/dp#144]
Going through  
https://github.com/nicholas-leonard/dp/blob/master/examples/neuralnetwork.lua
and stepping into the functions with a debugger I discovered the file 'optimizer.lua' which seems
to contain optimization of the model. However I was not able to find the exact place in the code where optimization is taking place. I was expecting that at some point in the code an there would be a call to an optimization method. Do I miss something?

**Comment by nicholas-leonard** [comment]
Its optimization in the sense of training a model. So when it calls forward/backward/updateParameters, then it is optimizing the model on a dataset.

**Comment by nicholas-leonard** [comment]
Propagator:backward calls model:backward (https://github.com/nicholas-leonard/dp/blob/master/propagator/optimizer.lua#L59). Also, the `self._callback` is a function defined in the neuralnetwork.lua script (user defined) for updating the model after each forward/backward.

**Comment by YitzhakSp** [comment]
As far as I understand model:backward computes the gradient. but where is the actual optimization?

**Comment by nicholas-leonard** [comment]
Yeah so the entire network (structure of modules) and criterion form a cost function. This cost function is minimized through gradient descent of the cost function w.r.t. the parameters. The chain rule is implemented through the backward call. The actual gradient descent (params = params - gradient), is performed via the updateParameters call. Which can be found in the user defined `callback` function : https://github.com/nicholas-leonard/dp/blob/master/examples/neuralnetwork.lua#L94 .
